# <font color=#0404B4>**SEMINAR - PHÂN TÍCH DỮ LIỆU THÔNG MINH** </font>

### <font color=#0404B4> Bảng thành viên nhóm </font>
|<center><div style="width:150px">MSSV</div><center>|<center><div style="width:290px">Họ và tên</div><center>|
|---------- |:-------------:|
| <center>20120032<center>  | <center>Phan Trường An<center> |
| <center>20120037<center>  | <center>Trần Thị Minh Anh<center> |
| <center>20120041<center>  | <center>Trần Kim Bảo<center> |
| <center>20120053<center>  |   <center>Nguyễn Thành Đạt<center>   | 
| <center>20120071<center>  |<center>Nguyễn Thị Bích Hà<center> |
| <center>20120095<center>  | <center>Ngô Quang Hưng<center> |

# <font color=#0404B4>Import thư viện</font>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set_theme()

<a name="2"></a>
# <font color=#0404B4> A. Tiền xử lý dữ liệu</font>


### 1. Đọc dữ liệu

In [2]:
item_category_df=pd.read_csv("item_categories.csv")
item_category_df.shape

(84, 2)

In [3]:
items_df=pd.read_csv("items.csv")
items_df.shape

(22170, 3)

In [4]:
sales_train_df=pd.read_csv("sales_train.csv")
sales_train_df.shape

(2935849, 6)

In [5]:
sample_submission_df=pd.read_csv("sample_submission.csv")
sample_submission_df.shape

(214200, 2)

In [6]:
shops_df=pd.read_csv("shops.csv")
shops_df.shape

(60, 2)

In [7]:
test_df=pd.read_csv("test.csv")
test_df.shape

(214200, 3)

### 2. Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?
- `item_category_df`: Mỗi dòng là thông tin về danh mục sản phẩm gồm: tên danh mục sản phẩm và id của danh mục sản phẩm đó.
- `items_df`: Mỗi dòng là thông tin về sản phẩm gồm: tên sản phẩm, id của sản phẩm và id của danh mục chứa sản phẩm đó.
- `sales_train_df`: Mỗi dòng là thông tin về bán hàng gồm: ngày bán hàng, thuộc kì bán nào(1 kì ở đây tương ứng với 1 tháng), id của cửa hàng bán sản phẩm đó, id của sản phẩm bán, thông tin về giá của sản phẩm và số lượng bán sản phẩm đó trong 1 ngày.
- `sample_submission_df`: Hướng dẫn định dạng nộp bài gồm các thông tin: ID(thứ tự) dự đoán, số lượng sản phẩm bán ra trong tháng.
- `shops_df`: Mỗi dòng là thông tin về cửa hàng gồm: tên cửa hàng và id của cửa hàng đó
- `test_df`: Chứa dữ liệu kiểm tra cho mô hình dự đoán gồm ID dự đoán, id cửa hàng và id sản phẩm.

### 3. Có dữ liệu bị trùng lặp ở các dataframe không

In [8]:
def check_duplicated(dataframe):
    duplicated_rows = dataframe.duplicated()

    # In các hàng bị lặp lại
    if len(dataframe[duplicated_rows])>0:
        print('Có dòng dữ liệu bị lặp')
        display(dataframe[duplicated_rows])
    else: 
        print('Không có dòng dữ liệu nào bị lặp')


Dataframe `item_category_df`

In [9]:
check_duplicated(item_category_df)

Không có dòng dữ liệu nào bị lặp


Dataframe `items_df`

In [10]:
check_duplicated(items_df)

Không có dòng dữ liệu nào bị lặp


Dataframe `sales_train_df`

In [11]:
check_duplicated(sales_train_df)

Có dòng dữ liệu bị lặp


date  date_block_num  shop_id  item_id  item_price  \
76962    05.01.2013               0       54    20130       149.0   
1435367  23.02.2014              13       50     3423       999.0   
1496766  23.03.2014              14       21     3423       999.0   
1671873  01.05.2014              16       50     3423       999.0   
1866340  12.07.2014              18       25     3423       999.0   
2198566  31.12.2014              23       42    21619       499.0   

         item_cnt_day  
76962             1.0  
1435367           1.0  
1496766           1.0  
1671873           1.0  
1866340           1.0  
2198566           1.0

Xóa luôn các mẫu dữ liệu bị trùng

In [12]:
sales_train_df.drop_duplicates(inplace=True)

Dataframe `sample_submission_df`

In [13]:
check_duplicated(sample_submission_df)

Không có dòng dữ liệu nào bị lặp


Dataframe `shops_df`

In [14]:
check_duplicated(shops_df)

Không có dòng dữ liệu nào bị lặp


Dataframe `test_df`

In [15]:
check_duplicated(test_df)

Không có dòng dữ liệu nào bị lặp


<h3> 4. Mỗi cột có ý nghĩa gì?</h3>

|STT|Tên cột dữ liệu | <center>Mô tả<center> | Đơn vị |
|--|:------:| ---| --- |
|1|item_category_name|<center>Tên danh mục sản phẩm<center>|
|2|item_category_id| <center>Mã định danh của danh mục sản phẩm<center>|
|3|item_name|<center>Tên sản phẩm<center>| 
|4|item_id| <center>Mã định danh sản phẩm<center>||
|5|date|<center>Ngày bán hàng<center>|| 
|6|date_block_num|<center>Thứ tự kỳ bán hàng *(bắt đầu từ 0)*<center>|
|7|shop_id|<center>Mã định danh cửa hàng<center>| |
|8|item_price|<center>Giá sản phẩm<center>|<center>Rúp<center>|
|9|item_cnt_day|<center>Số lượng bán ra/ số lượng mất trong `1 ngày`<center>|<center>Sản phẩm <center>|
|10|ID|<center>Mã định danh<center>| |
|11|item_cnt_month|<center>Số lượng bán ra/ số lượng mất trong `1 tháng`<center>|<center>Sản phẩm <center>|
|12|shop_name|<center>Tên cửa hàng<center>| |


### 5. Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp?



Dataframe `item_category_df`

In [16]:
item_category_df.columns

Index(['item_category_name', 'item_category_id'], dtype='object')

In [17]:
item_category_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84 entries, 0 to 83
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   item_category_name  84 non-null     object
 1   item_category_id    84 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.4+ KB


Dataframe `items_df`

In [18]:
items_df.columns

Index(['item_name', 'item_id', 'item_category_id'], dtype='object')

In [19]:
items_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22170 entries, 0 to 22169
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   item_name         22170 non-null  object
 1   item_id           22170 non-null  int64 
 2   item_category_id  22170 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 519.7+ KB


Dataframe `sales_train_df`

In [20]:
sales_train_df.columns

Index(['date', 'date_block_num', 'shop_id', 'item_id', 'item_price',
       'item_cnt_day'],
      dtype='object')

In [21]:
sales_train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2935843 entries, 0 to 2935848
Data columns (total 6 columns):
 #   Column          Dtype  
---  ------          -----  
 0   date            object 
 1   date_block_num  int64  
 2   shop_id         int64  
 3   item_id         int64  
 4   item_price      float64
 5   item_cnt_day    float64
dtypes: float64(2), int64(3), object(1)
memory usage: 156.8+ MB


Dataframe `sample_submission_df`

In [22]:
sample_submission_df.columns

Index(['ID', 'item_cnt_month'], dtype='object')

In [23]:
sample_submission_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214200 entries, 0 to 214199
Data columns (total 2 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   ID              214200 non-null  int64  
 1   item_cnt_month  214200 non-null  float64
dtypes: float64(1), int64(1)
memory usage: 3.3 MB


Dataframe `shops_df`

In [24]:
shops_df.columns

Index(['shop_name', 'shop_id'], dtype='object')

In [25]:
shops_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60 entries, 0 to 59
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   shop_name  60 non-null     object
 1   shop_id    60 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 1.1+ KB


Dataframe `test_df`

In [26]:
test_df.columns

Index(['ID', 'shop_id', 'item_id'], dtype='object')

In [27]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 214200 entries, 0 to 214199
Data columns (total 3 columns):
 #   Column   Non-Null Count   Dtype
---  ------   --------------   -----
 0   ID       214200 non-null  int64
 1   shop_id  214200 non-null  int64
 2   item_id  214200 non-null  int64
dtypes: int64(3)
memory usage: 4.9 MB



##### Cột `date` của dataframe `sales_train_df` đang có kiểu dữ liệu `object`. Thực hiện xử lý đưa về đúng kiểu dữ liệu `datime64[ns]`.

In [28]:
sales_train_df.date=pd.to_datetime(sales_train_df.date,format="%d.%m.%Y")

In [29]:
sales_train_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2935843 entries, 0 to 2935848
Data columns (total 6 columns):
 #   Column          Dtype         
---  ------          -----         
 0   date            datetime64[ns]
 1   date_block_num  int64         
 2   shop_id         int64         
 3   item_id         int64         
 4   item_price      float64       
 5   item_cnt_day    float64       
dtypes: datetime64[ns](1), float64(2), int64(3)
memory usage: 156.8 MB



### 7. Với mỗi cột, các giá trị được phân bố như thế nào?

- Nhận xét:
    - Các dataframe như: `item_category_df`, `items_df`, `sample_submission_df`, `shops_df` và `test_df`, là những dataframe chứa thông tin chi tiết đối đối với dataframe `sales_train_df`. Nên sẽ chỉ tập trung xem phân bố dữ liệu của dataframe `sales_train_df`
    - Với các dữ liệu trong dataframe `sales_train_df` chỉ có dạng số.

#### Cột có kiểu dữ liệu dạng số, và phân bố:


In [30]:
numerical_cols = sales_train_df[sales_train_df.select_dtypes(exclude=['object']).columns]

def missing_ratio(c):
    return  c.isna().sum()/len(c) * 100

def min(c):
    return c.min()

def max(c):
    return c.max()

nume_col_profiles_df = numerical_cols.agg([missing_ratio, min, max])
nume_col_profiles_df

date  date_block_num  shop_id  item_id  \
missing_ratio                  0.0             0.0      0.0      0.0   
min            2013-01-01 00:00:00             0.0      0.0      0.0   
max            2015-10-31 00:00:00            33.0     59.0  22169.0   

               item_price  item_cnt_day  
missing_ratio         0.0           0.0  
min                  -1.0         -22.0  
max              307980.0        2169.0

- Nhận xét:
    - `date`: dữ liệu bán hàng được thu thập từ ngày 01-01-2013 đến ngày 31-10-2015.
    - `date_block_num`: Chia dữ liệu thành 34 kì (1 kì=1 tháng)
    - `shop_id`: ID cửa hàng có giá trị nhỏ nhất là 0, lớn nhất là 59.
    - `item_id`: ID sản phẩm có giá trị nhỏ nhất là 0, lớn nhất là 22169.
    - `item_price`: Sản phẩm có giá tiền nhỏ nhất -1, lớn nhất là 307980 (Rúp).
    - `item_cnt_day`: Số lượng sản phẩm đó được bán ra. Số lượng sản phẩm bị mất lớn nhất là 22 sản phẩm, có thể nhiều lý do như hư hỏng, thất lạc trong quá trình vận chuyển,... Số lượng sản phẩm được bán ra lớn nhất là 2169 sản phẩm. 

###  8. Có cột nào cần phải tiền xử lý hay không?


Kiểm tra thử số lượng `item_price` có giá bị âm

In [31]:
sales_train_df[sales_train_df.item_price<0]

date  date_block_num  shop_id  item_id  item_price  item_cnt_day
484683 2013-05-15               4       32     2973        -1.0           1.0

Với chỉ 1 giá sản phẩm bị âm nên ta có thể thực hiện loại xóa dữ liệu đó.

In [32]:
index_drop=sales_train_df[sales_train_df.item_price<0].index
sales_train_df.drop(index_drop)

date  date_block_num  shop_id  item_id  item_price  item_cnt_day
0       2013-01-02               0       59    22154      999.00           1.0
1       2013-01-03               0       25     2552      899.00           1.0
2       2013-01-05               0       25     2552      899.00          -1.0
3       2013-01-06               0       25     2554     1709.05           1.0
4       2013-01-15               0       25     2555     1099.00           1.0
...            ...             ...      ...      ...         ...           ...
2935844 2015-10-10              33       25     7409      299.00           1.0
2935845 2015-10-09              33       25     7460      299.00           1.0
2935846 2015-10-14              33       25     7459      349.00           1.0
2935847 2015-10-22              33       25     7440      299.00           1.0
2935848 2015-10-03              33       25     7460      299.00           1.0

[2935842 rows x 6 columns]

<a name="2"></a>
# <font color=#0404B4> B. Khám phá dữ liệu</font>


### Tìm ra tháng có tiền từ doanh số bán hàng cao nhất. Phân tích tương quan giữa số tiền mỗi shop kiếm được với số mặt hàng các shop đã bán.

#### Bước 1: Tiến hành gom nhóm và tính tổng số tiền bán được của tất cả các shop.

In [46]:
df_income=sales_train_df.copy()
df_income['month']=df_income['date'].dt.month
df_income['year']=df_income['date'].dt.year.astype(str)
df_income['income']=df_income['item_cnt_day']*df_income['item_price']
df_income=df_income.groupby(['year','month']).sum()
df_income=df_income.reset_index()[['year','month','income']]
df_income.head()

year  month        income
0  2013      1  9.194694e+07
1  2013      2  9.066571e+07
2  2013      3  1.049327e+08
3  2013      4  6.915429e+07
4  2013      5  6.506531e+07

#### Bước 2: Trực quan bằng `line chart`.

- Dùng `line chart` có thể giúp quan sát rõ biến động và xu hướng tăng của nguồn thu nhập của các shop.

In [50]:
import plotly.graph_objs as go
    # Tạo đối tượng Figure và Axes
plt.rcParams['figure.dpi'] = 360


df_income['month']=df_income['month'].astype(str)
df_income['year']=df_income['year'].astype(str)

df_income['month_year']=df_income['month']+'-'+df_income['year']

length=len(df_income['month_year'])
color_markers=[]
for i in range(length):
    color_markers.append('#FF0499')

# Tạo đối tượng đồ thị tương tác
fig = go.Figure()

# Thêm dữ liệu vào đồ thị
fig.add_trace(go.Scatter(x=df_income['month_year'], y=df_income['income'], mode='markers+lines', name='line',
                            marker=dict(size=10, color=color_markers)))

# Cấu hình đồ thị
fig.update_layout(title='Biểu đồ biểu diễn tổng thu nhập của tất cả shop qua từng tháng',
                yaxis_title='Số tiền',
                xaxis_title='Tháng-Năm')

# Hiển thị biểu đồ
fig.show()

- Nhận xét: 
    + Tổng thu nhập của các shop sẽ có xu hướng giảm ở giữa năm, từ tháng 4 đến tháng 7. Vì thời điểm này thường là mùa xuân và mùa hè đầu, nhưng vẫn còn khá lạnh và mưa ở nhiều khu vực, khiến người dân không có nhu cầu mua sắm đồ dùng cho mùa hè. Bên cạnh đó, trong thời gian này, không có các ngày lễ lớn hoặc sự kiện quan trọng nào.
    + Tổng thu nhập các shop tăng cao vào cuối năm. Có thể thấy, trong biểu đồ, đường biểu diễn giá trị đạt đỉnh ở các tháng 12. Lý giải là vì, Ở Nga, tháng 12 được coi là một tháng đặc biệt vì nó gần Giáng sinh và Lễ Tạ ơn Chúa. Do đó, vào tháng 12, nhu cầu mua sắm tăng cao khi mọi người chuẩn bị cho các ngày lễ này. Ngoài ra, tháng 12 cũng là tháng cuối cùng của năm, và nhiều người sẽ dành thời gian để mua sắm đồ để tặng người thân và bạn bè trong dịp kỷ niệm năm mới.
    > Chọn tháng 12 để phân tích tương quan các trường dữ liệu.

#### Bước 3: Lọc những hoạt động mua bán vào tháng 12 và tiến hành gom nhóm theo `shop_id` để tính số lượng mặt hàng đã bán của mỗi shop.

In [55]:
df_temp=sales_train_df.copy()
df_temp=df_temp.loc[df_temp['date'].dt.month==12]
grouped_shop=df_temp.groupby('shop_id')
grouped_shop=grouped_shop.count().reset_index()[['shop_id','item_id']].sort_values('item_id',ascending=False)
grouped_shop.head()

shop_id  item_id
24       31    21987
18       25    17381
45       54    13069
48       57    11834
21       28    11516

#### Bước 4: Trực quan bằng `Tree map`.

- `Tree map` rất phù hợp để trực quan cho dữ liệu có nhiều loại phân biệt khác nhau.
> Dùng `tree map` để có thể quan sát rõ hơn nhiều shop và số lượng mặt hàng của mỗi shop bán được.

In [53]:
plt.rcParams['figure.dpi'] = 360

import plotly.express as px

# Tạo Treemap
fig = px.treemap(grouped_shop, path=['shop_id'], values='item_id', color='shop_id',
                 hover_data=['shop_id'], color_discrete_sequence=px.colors.qualitative.D3_r,
                 width=900, height=600, title='Biểu đồ biểu diễn số lượng mặt hàng của các shop vào tháng 12')

# Cấu hình thông tin hiển thị
fig.update_traces(textinfo='label+value', hovertemplate='<b>%{label}</b><br>Value: %{value}')


# Hiển thị biểu đồ
fig.show()

#### Bước 5: Tiến hành gom nhóm `shop_id` để tính tổng thu nhập mỗi shop vào tháng 12.

In [56]:
grouped_shop_income=df_temp.groupby('shop_id').sum().reset_index()[['shop_id','item_price']]
grouped_shop_income['count_id']=grouped_shop['item_id']
grouped_shop_income.head()

shop_id    item_price  count_id
0        2  4.716513e+06      2688
1        3  2.968652e+06      2217
2        4  4.426936e+06      3614
3        5  4.227661e+06      3673
4        6  9.259795e+06      7927

#### Bước 6: Trực quan bằng `bubble chart`.
- `Bubble chart` rất phù hợp để xét độ lớn giữa các dữ liệu và đưa dữ liệu các dữ liệu có giá trị quy định lớn nổi lên.
> Dùng `bubble chart` để quan sát các shop nào có thu nhập cao để tương quan với `tree map` biểu diễn số lượng mặt hàng của các shop.

In [57]:
import plotly.express as px

# Tạo Bubble chart
fig = px.scatter(grouped_shop_income, x='shop_id', y='count_id', size='item_price', color='shop_id',
                 hover_data=['shop_id'], text='shop_id')

# Cấu hình tên và tiêu đề trục
fig.update_xaxes(title_text='Shop ID')
fig.update_yaxes(title_text='Số lượng sản phẩm')

# Hiển thị biểu đồ
fig.show()
    


- Nhận xét: 
    - Các shop có thu nhập trên 10,000 chỉ tính ở tháng 12 có id là: 31, 25, 54, 57, 28, 27, 42 (theo thứ tự giảm dần).
    - Hầu hết các shop này đều có trên 10,000 mặt hàng khác nhau được bán.

> Tương quan 2 biểu đồ, ta thấy các shop có số lượng mặt hàng cao nhất cũng theo thứ tự giảm dần đều tương ứng các shop có số tiền thu nhập cao nhất vào tháng 12. 
    > Ta có thể thấy, shop càng có đa dạng số lượng mặt hàng bán được thì sẽ có số tiền bán được càng cao.